### PR TITLE
Improve opaque type with no RHS error message

### LINKED ISSUE #15266: Misleading error message for opaque type
## Compiler version

3.1.2

## Minimized example

```Scala
opaque type Type
```

## Output

```scala
[error] Modifier opaque is not allowed for this definition
[error] opaque type Type
[error] ^^^^^^^^^^^^^^^^
```

## Expectation

Should state that an opaque type needs a right-hand side

### PR DESCRIPTION
**Pull request by nicolasstucki**
Fixes #15266

Done in the spree with @griggt and @mtk

### CODE REVIEW

**Comment by griggt** [comment]
It turns out this isn't the first time we've had issues with missing RHS of an opaque type, we have `tests/neg/i6055.scala`:

```scala
opaque type i0           // error: opaque type must be an alias type
opaque type i2 <: Int    // error: opaque type must be an alias type

opaque type i1[_]        // error: opaque type must be an alias type
opaque type x[_] <: Int  // error: opaque type must be an alias type
```

At one point the error message was as shown above, but this test has no check file, so the loss in clarity may not have been noticed at the time it was introduced.

More relevant for this PR, the new output for the above test is

```scala
-- Error: tests/neg/i6055.scala:1:12 -------------------------------------------
1 |opaque type i0           // error: opaque type must be an alias type
  |^^^^^^^^^^^^^^
  |opaque type must have a right-hand side
-- Error: tests/neg/i6055.scala:2:12 -------------------------------------------
2 |opaque type i2 <: Int    // error: opaque type must be an alias type
  |^^^^^^^^^^^^^^^^^^^^^
  |opaque type must have a right-hand side
-- [E156] Syntax Error: tests/neg/i6055.scala:4:12 -----------------------------
4 |opaque type i1[_]        // error: opaque type must be an alias type
  |^^^^^^^^^^^^^^^^^
  |Modifier opaque is not allowed for this definition
-- [E156] Syntax Error: tests/neg/i6055.scala:5:12 -----------------------------
5 |opaque type x[_] <: Int  // error: opaque type must be an alias type
  |^^^^^^^^^^^^^^^^^^^^^^^
  |Modifier opaque is not allowed for this definition
```

We should enhance the checking to recurse on LambdaTypeTree nodes so we detect the missing RHS on higher-kinded opaque types.

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala [hunk 1]
```diff
@@ -1235,18 +1235,27 @@ object desugar {
       rhsOK(rhs)
   }
 
+  def checkOpaqueAlias(tree: MemberDef)(using Context): MemberDef =
+    if !tree.mods.is(Opaque) then tree
+    else tree match
+      case TypeDef(_, bounds: TypeBoundsTree) if bounds.alias.isEmpty =>
+        report.error(i"opaque type must have a right-hand side", tree.srcPos)
```
Comment (griggt) [inline compiler/src/dotty/tools/dotc/ast/Desugar.scala:1242]:
```suggestion
        report.error(i"opaque type must have a right-hand side", tree.srcPos.endPos)
```

...if we want the caret to be placed where the RHS should go?

**Comment by griggt** [comment]
> We should enhance the checking to recurse on LambdaTypeTree nodes so we detect the missing RHS on higher-kinded opaque types.

I pushed a commit to do this.

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/ast/Desugar.scala:1241]: Both approaches seem reasonable to me. I do not have a preference on either option.

**Review by smarter — APPROVED** [review]

**Pull request merged**